File: docker/mcp-server/src/tools/vectortools.js
const { z } = require('zod');
const { apiGet, apiPost } = require('../api');

// The XSS scanners: what each one can actually reach, how it is configured, and what a run found.
//
// SETTINGS ARE THE SAME STORE THE UI WRITES. A value changed here appears in the Config modal
// immediately and the other way round, because there is one copy on the server rather than a UI copy
// and an agent copy. That is also why there is no local list of flags in this file: option_reference
// returns the server's own vocabulary, which is the same map the command composer reads.
//
// THE ONE THING TO UNDERSTAND BEFORE RUNNING ANYTHING: only Dalfox can reach every insertion point.
// domdig fuzzes the query string and the URL hash; its cookie and header flags set STATIC values for
// authentication and it navigates rather than issuing a POST. xssFuzz calls requests.get and nothing
// else. Handed a header, body, cookie or path vector, either one scans the query string, finds
// nothing and exits 0, so "no findings" from them is not evidence about those vectors. eligibility
// reports how many of the target's vectors each tool can test, and results lists the untested ones
// with the reason, which is why both exist.

const buildSchema = (tools) => z.object({
  action: z.enum(['eligibility', 'option_reference', 'settings', 'save_settings',
    'section_settings', 'save_section_settings',
    'run', 'status', 'results', 'triage']).describe(
    'eligibility: how many of this target\'s attack vectors each tool can actually test, which ' +
    'insertion points it can reach, and which settings are currently blinding it. START HERE, ' +
    'because two of the three tools cover only query parameters and report clean for everything ' +
    'else. ' +
    'option_reference: every setting a tool honours, with its flag, type and what the tool does ' +
    'when it is not set. This is the server\'s own vocabulary, so it never drifts from what a scan ' +
    'reads. ' +
    'settings: the stored settings for one tool on this target. ' +
    'save_settings: MERGE settings into the store. Send a key as null to remove it. Pass ' +
    'replace: true to make the payload authoritative. Refused if it names a flag the runner owns. ' +
    'section_settings: the settings that belong to the SECTION rather than to any one tool, which ' +
    'is where the out-of-band webhook pair lives. A tool that requires one reports every vector as ' +
    'skipped until it is set, so check here first when eligibility is 0 of everything. ' +
    'save_section_settings: MERGE into that section store. Nuclei DAST and REcollapse need BOTH ' +
    'listeningWebhookURL (where the target calls out to) and resultsWebhookURL (the API the runner ' +
    'reads the hits back from); one without the other counts as unconfigured. ' +
    'run: scan every eligible attack vector with one tool, one vector at a time. ' +
    'status: how a run is going, including which vector of how many. ' +
    'results: the findings AND the vectors that were never sent, with the reason for each. ' +
    'triage: mark one finding new, interesting or dismissed.'),
  target_id: z.string().optional().describe('Scope target id. Defaults to the active target.'),
  tool: z.enum(tools).optional().describe(
    'Which scanner. Required for everything except eligibility, which covers all of them when omitted.'),
  settings: z.record(z.any()).optional().describe(
    'Settings to store, keyed by the option keys option_reference lists. A null value removes a key.'),
  replace: z.boolean().optional().describe(
    'Make the settings payload authoritative rather than merging it into what is stored.'),
  finding_id: z.string().optional().describe('Finding to triage.'),
  triage: z.enum(['new', 'interesting', 'dismissed']).optional(),
});



async function resolveTargetId(params) {
  if (params.target_id) return params.target_id;
  const targets = await apiGet('/scopetarget/read');
  const active = (Array.isArray(targets) ? targets : []).find((t) => t.active);
  return active ? active.id : null;
}

async function manageVectorTools(category, tools, params) {
  const targetId = await resolveTargetId(params);
  if (!targetId && params.action !== 'option_reference' && params.action !== 'triage') {
    return { error: 'No target_id given and no active target set.' };
  }

  switch (params.action) {
    case 'option_reference': {
      const registry = await apiGet(`/${category}/tools`);
      const tools = (registry.tools || []).filter((t) => !params.tool || t.key === params.tool);
      return {
        tools: tools.map((t) => ({
          tool: t.key,
          name: t.name,
          insertion_points_it_can_reach: t.insertion_points,
          limitation: t.limitation || undefined,
          groups: t.groups,
          options: t.options,
          // Named so an agent does not waste a round trip trying to set one and getting a 400.
          flags_the_framework_sets: t.owned_flags,
        })),
      };
    }

    case 'eligibility': {
      const wanted = params.tool ? [params.tool] : tools;
      const reports = await Promise.all(
        wanted.map((tool) => apiGet(`/${category}/${targetId}/${tool}/status`).catch(() => null)),
      );
      return {
        tools: reports.filter(Boolean).map((r) => ({
          tool: r.tool,
          eligible: r.eligibility?.eligible ?? 0,
          total: r.eligibility?.total ?? 0,
          can_reach: r.eligibility?.reachable || [],
          cannot_reach: r.eligibility?.unreachable || [],
          skipped_by_insertion_point: r.eligibility?.skipped_by_point || {},
          blinded_by_current_settings: r.eligibility?.blinded || {},
          limitation: r.eligibility?.limitation || undefined,
        })),
        note: 'cannot_reach vectors are never sent. A tool reporting no findings says nothing '
          + 'about them.',
      };
    }

    case 'settings': {
      if (!params.tool) return { error: 'tool is required for settings.' };
      return apiGet(`/${category}/${targetId}/${params.tool}/settings`);
    }

    case 'save_settings': {
      if (!params.tool) return { error: 'tool is required for save_settings.' };
      if (!params.settings) return { error: 'settings is required for save_settings.' };
      return apiPost(`/${category}/${targetId}/${params.tool}/settings`, {
        settings: params.settings,
        replace: params.replace === true,
      });
    }

    // Section settings belong to the SECTION, not to a tool, so they take no tool argument. They
    // were unreachable from here until now: nuclei-dast and REcollapse both declare
    // RequiresSectionSetting: "listeningWebhookURL", so with no way to set it every vector was
    // reported skipped and eligibility read 0 of 53 with nothing an agent could do about it.
    case 'section_settings':
      return apiGet(`/${category}/${targetId}/section-settings`);

    case 'save_section_settings': {
      if (!params.settings) return { error: 'settings is required for save_section_settings.' };
      // The server REPLACES this store wholesale, so a merge has to happen here or setting one half
      // of the webhook pair would silently delete the other half.
      let merged = params.settings;
      if (params.replace !== true) {
        const current = await apiGet(`/${category}/${targetId}/section-settings`).catch(() => null);
        merged = { ...((current && current.settings) || {}), ...params.settings };
      }
      for (const [key, value] of Object.entries(merged)) {
        if (value === null) delete merged[key];
      }
      return apiPost(`/${category}/${targetId}/section-settings`, { settings: merged });
    }

    case 'run': {
      if (!params.tool) return { error: 'tool is required for run.' };
      return apiPost(`/${category}/${targetId}/${params.tool}/scan`, {});
    }

    case 'status': {
      if (!params.tool) return { error: 'tool is required for status.' };
      return apiGet(`/${category}/${targetId}/${params.tool}/status`);
    }

    case 'results': {
      if (!params.tool) return { error: 'tool is required for results.' };
      const data = await apiGet(`/${category}/${targetId}/${params.tool}/results`);
      return {
        ...data,
        note: data.skipped && data.skipped.length
          ? `${data.skipped.length} attack vectors were never sent to ${params.tool}. They are `
            + 'unknown, not clean. Each carries the reason in skipped[].reason.'
          : undefined,
      };
    }

    case 'triage': {
      if (!params.finding_id) return { error: 'finding_id is required for triage.' };
      if (!params.triage) return { error: 'triage is required.' };
      return apiPost(`/${category}/finding/${params.finding_id}/triage`, { triage: params.triage });
    }

    default:
      return { error: `unknown action: ${params.action}` };
  }
}

// One pair per section. The schema differs only in which tool names are valid, and the handler is
// the same code, because the API routes are the same shape under a different prefix.
const XSS_TOOLS = ['dalfox', 'domdig', 'xssfuzz'];
const SQLI_TOOLS = ['sqlmap', 'ghauri', 'sqlidetector'];
const CACHE_TOOLS = ['wcvs', 'cacheboom'];
const CMDI_TOOLS = ['commix', 'sstimap', 'tinja'];
const REDIRECT_TOOLS = ['nuclei-dast', 'recollapse', 'ssrfmap'];
const LFI_TOOLS = ['lfimap', 'lfihunt'];
const SMUGGLING_TOOLS = ['smugglex', 'http2smugl'];
const BYPASS_TOOLS = ['nomore403', 'forbidden'];
const GRAPHQL_TOOLS = ['graphql-cop', 'clairvoyance', 'graphw00f'];
const LEAK_TOOLS = ['snallygaster', 'mantra', 'trufflehog'];
const GIT_TOOLS = ['git-dumper', 'gittools'];
const MISC_TOOLS = ['upload-bypass', 'jwt-tool', 'pphack'];

const manageXSSSchema = buildSchema(XSS_TOOLS);
const manageSQLiSchema = buildSchema(SQLI_TOOLS);
const manageCacheSchema = buildSchema(CACHE_TOOLS);
const manageCmdiSchema = buildSchema(CMDI_TOOLS);
const manageRedirectSchema = buildSchema(REDIRECT_TOOLS);
const manageLfiSchema = buildSchema(LFI_TOOLS);
const manageSmugglingSchema = buildSchema(SMUGGLING_TOOLS);
const manageBypassSchema = buildSchema(BYPASS_TOOLS);
const manageGraphqlSchema = buildSchema(GRAPHQL_TOOLS);
const manageLeakSchema = buildSchema(LEAK_TOOLS);
const manageGitSchema = buildSchema(GIT_TOOLS);
const manageMiscSchema = buildSchema(MISC_TOOLS);

const manageXSS = (params) => manageVectorTools('xss', XSS_TOOLS, params);
const manageSQLi = (params) => manageVectorTools('sqli', SQLI_TOOLS, params);
const manageCache = (params) => manageVectorTools('cache', CACHE_TOOLS, params);
const manageCmdi = (params) => manageVectorTools('cmdi', CMDI_TOOLS, params);
const manageRedirect = (params) => manageVectorTools('redirect-ssrf', REDIRECT_TOOLS, params);
const manageLfi = (params) => manageVectorTools('lfi', LFI_TOOLS, params);
const manageSmuggling = (params) => manageVectorTools('smuggling', SMUGGLING_TOOLS, params);
const manageBypass = (params) => manageVectorTools('access-bypass', BYPASS_TOOLS, params);
const manageGraphql = (params) => manageVectorTools('graphql', GRAPHQL_TOOLS, params);
const manageLeak = (params) => manageVectorTools('sensitive-leak', LEAK_TOOLS, params);
const manageGit = (params) => manageVectorTools('exposed-git', GIT_TOOLS, params);
const manageMisc = (params) => manageVectorTools('misc', MISC_TOOLS, params);

module.exports = {
  manageXSSSchema, manageXSS,
  manageSQLiSchema, manageSQLi,
  manageCacheSchema, manageCache,
  manageCmdiSchema, manageCmdi,
  manageRedirectSchema, manageRedirect,
  manageLfiSchema, manageLfi,
  manageSmugglingSchema, manageSmuggling,
  manageBypassSchema, manageBypass,
  manageGraphqlSchema, manageGraphql,
  manageLeakSchema, manageLeak,
  manageGitSchema, manageGit,
  manageMiscSchema, manageMisc,
};

File: server/utils/vectorTools.go
package utils

import (
	"context"
	"sort"
	"strings"
	"time"
)

// The tools that TEST attack vectors, as one registry.
//
// XSS came first and SQL injection turned out to be the same shape: a set of scanners, each able to
// reach some subset of the five insertion points, each configured from a server-side store that the
// Config modal and the MCP tool both edit. Ten more sections are planned. Copying the settings
// store, the eligibility check, the runner and the API once per section would be roughly nine
// hundred lines duplicated twelve times, and the copies only have to disagree once for a scan to
// behave unlike the screen that configured it.
//
// So everything generic lives here and in vectorSettings.go, vectorEligibility.go, vectorScan.go and
// vectorAPI.go. What stays per tool is the part that genuinely differs: its option vocabulary, how a
// command line is built from a vector, and how its output is read. Those arrive as Compose and Parse
// on the struct below, which is the whole extension point.

// VectorOptionMeta is one setting: enough for a form field to be built from it and for a command
// line to be composed from it, without either side hardcoding the other's knowledge.
type VectorOptionMeta struct {
	Kind    string   `json:"kind"` // bool | int | float | string | enum | csv | path
	Group   string   `json:"group"`
	Label   string   `json:"label"`
	Flag    string   `json:"flag,omitempty"`
	Choices []string `json:"choices,omitempty"`
	// Placeholder says what the TOOL does when the option is not set, so an empty field reads as a
	// decision rather than as a gap.
	Placeholder string `json:"placeholder,omitempty"`
	// Repeatable options are passed once per value rather than comma joined. dalfox's -H and sqlmap's
	// --tamper differ here, and comma joining a header produces one header whose value has a comma.
	Repeatable bool `json:"repeatable,omitempty"`
}

// VectorFinding is one result, normalised across tools that agree on nothing.
type VectorFinding struct {
	VectorID string
	Tool     string
	// Kind is the tool's OWN confidence class, kept rather than flattened. dalfox's V and A, and
	// sqlmap's boolean-based versus error-based, mean genuinely different things, and collapsing them
	// would put an informational note beside a proven injection.
	Kind            string
	Severity        string
	Confidence      string
	InsertionPoint  string
	Param           string
	Payload         string
	Method          string
	URL             string
	Evidence        string
	DetectionMethod string
	InjectType      string
	RawRequest      string
	RawResponse     string
	// IsBypassTarget says whether VectorID names an attack vector or an access bypass target. The two
	// live in different tables with different foreign keys.
	IsBypassTarget bool
	// IsGraphQLTarget says the id names an endpoint from a tool's own settings list, which exists in
	// no table at all, so it is stored against neither foreign key.
	IsGraphQLTarget bool
	// IsLeakTarget marks a directory from the sensitive data leak list, which has its own table.
	IsLeakTarget bool
}

// VectorComposer turns one attack vector plus the stored settings into one command line. Returning
// nil args means the tool REFUSED this vector, and the warnings say why.
type VectorComposer func(v VectorInput, settings map[string]any, reportPath string) ([]string, []string)

// VectorParser reads whatever the tool produced. Given both the captured stdout and the contents of
// the report file, because the three XSS tools and the three SQL tools between them use both.
type VectorParser func(stdout, report string, row vectorRow) []VectorFinding

// VectorTool is one scanner: what it can reach, how it is configured, and how it is driven.
type VectorTool struct {
	Key             string                      `json:"key"`
	Name            string                      `json:"name"`
	Category        string                      `json:"category"`
	Binary          string                      `json:"binary"`
	Container       string                      `json:"container"`
	Groups          []string                    `json:"groups"`
	Options         map[string]VectorOptionMeta `json:"options"`
	OwnedFlags      map[string]string           `json:"owned_flags"`
	InsertionPoints []string                    `json:"insertion_points"`
	Limitation      string                      `json:"limitation,omitempty"`

	// Blinding names settings that make an insertion point untestable while the scan still exits 0
	// and reports nothing. Keyed by setting, valued by the point it blinds, or "all".
	Blinding map[string]string `json:"blinding,omitempty"`

	// ReportPath is set when the tool writes its findings to a file rather than to stdout. The runner
	// reads that file back out of the shared /tmp volume.
	UsesReportFile bool `json:"-"`

	Compose    VectorComposer      `json:"-"`
	Parse      VectorParser        `json:"-"`
	SkipReason func(string) string `json:"-"`

	// AttemptsWhenEmpty is how many times to run a FLAKY detector on one vector before believing it
	// found nothing. Findings stop the retries immediately, so a working tool pays nothing.
	//
	// pphack is why. It drives headless Chrome and watches whether the page's own JavaScript merges
	// the payload into Object.prototype, and whether that happens depends on script load timing.
	// Measured against ginandjuice.shop/blog, which is definitively vulnerable and which pphack
	// reports on demand from the command line: 3 hits in 6 identical runs. One run per vector is a
	// coin flip, and the framework ran each vector exactly once, so 24 vectors came back clean on a
	// target whose prototype pollution the tool finds in seconds when asked twice.
	//
	// Left at 0 for every deterministic tool, where a retry would only double the traffic.
	AttemptsWhenEmpty int `json:"-"`

	// Incomplete reports, from the tool's own output, that the run DID NOT FINISH, and why. A
	// non-empty return marks the vector as an error rather than clean, while any findings already
	// parsed are still kept.
	//
	// This exists because a tool can tell you it gave up and the runner can still file the result as
	// clean. dalfox aborted every one of 53 vectors on --on-session-loss abort, writing
	// {"meta":{"findings_count":0,"incomplete":true,...,"error_code":"SESSION_LOST"}}, and the parser
	// skips meta lines, so the reason was read and thrown away. The operator saw 53 clean vectors on
	// an application whose own /vulnerabilities page lists four separate XSS.
	Incomplete func(stdout, report string) string `json:"-"`

	// VectorEligible refuses an INDIVIDUAL vector, with a reason, where the insertion point alone is
	// not the whole story.
	//
	// Commix is why this exists. It reaches a header only at --level 3, and only for the three names
	// it knows: user-agent, referer and host. Measured, a Referer carrying the bug is found and a
	// custom X-Name carrying the same bug is not, even with commix's own INJECT_HERE marker. Saying
	// "commix can reach headers" would then be true and useless, because every header vector this
	// framework produces is a custom one.
	VectorEligible func(VectorInput) (bool, string) `json:"-"`

	// RequiresFinding gates a tool on what an earlier scan already found.
	//
	// The Open Redirect and SSRF section is why. REcollapse mutates a value to get past a validation
	// regex and SSRFmap pivots an SSRF into file read or worse; neither DETECTS anything, and both
	// are meaningless against a parameter that has not been shown to redirect or to fetch. Running
	// them across a whole vector table would be thousands of requests to learn nothing, and in
	// SSRFmap's case would aim internal port scans at a target on the strength of a guess.
	//
	// The value is the category whose findings count, so a tool is gated on another tool's work.
	RequiresFinding string `json:"requires_finding,omitempty"`

	// RequiresSectionSetting names a SECTION setting the tool cannot work without.
	//
	// The webhook is why. A payload that points at a callback URL nobody configured points at
	// nothing, and the scan then completes having proved only that an empty string is not an SSRF.
	// Reported as ineligible with the reason rather than run.
	RequiresSectionSetting string `json:"requires_section_setting,omitempty"`

	// DedupeKey collapses vectors that would produce the SAME scan.
	//
	// The cache tools are why this exists. They take a URL and discover unkeyed inputs from their own
	// wordlists; they do not test the parameter a vector names. Seventy-one vectors across thirty-odd
	// URLs would therefore run the same multi-thousand-request scan over and over against the same
	// URL, which is wasteful against our own infrastructure and abusive against someone else's. When
	// set, one representative per key is scanned and the findings are attributed to every vector that
	// shares it. Nil means one scan per vector, which is right for XSS and SQL injection because
	// there the vector IS the thing being tested.
	DedupeKey func(VectorInput) string `json:"-"`

	// ScanUnit names what a run counts, for a tool whose unit is neither a vector nor a URL.
	ScanUnit string `json:"-"`

	// RowSource replaces the attack vector table as the thing this tool scans.
	//
	// The access control bypass section is why. Its targets are the URLs that already answered 401 or
	// 403 to something an earlier tool sent, consolidated from six tables that record an HTTP status.
	// Those are not attack vectors and must not be put in that table: every other section loads it
	// wholesale, so a few hundred 403 URLs would quietly become XSS and SQL injection targets too.
	RowSource func(context.Context, string) ([]vectorRow, error) `json:"-"`

	// Runs are the sub-scans a single vector needs. CacheBoom's -m is required and takes one of cd or
	// cp, so covering both means running it twice. Empty means a single unlabelled run.
	Runs []string `json:"runs,omitempty"`

	// Timeout bounds one run. Zero means the default.
	Timeout time.Duration `json:"-"`

	// TimeoutIsNormal marks a tool that does not exit on its own. CacheBoom's deception mode blocks
	// forever on task_queue.join(), because its workers return on the found flag without calling
	// task_done() for the items still queued: it finds the bug, prints it, and never terminates.
	// Killing it is the expected end of a successful run, not an error to report.
	TimeoutIsNormal bool `json:"-"`
}

// VectorInsertionPoints is every insertion point an attack vector can carry. Kept here so an
// eligibility report can name the ones a tool CANNOT reach, not only the ones it can.
var VectorInsertionPoints = []string{"query", "body", "header", "cookie", "path"}

// VectorCanary is the value substituted for a parameter whose name is known but whose value was
// never observed. On the reference target 16 of 27 query vectors are in that state: discovered by
// Arjun, x8 or ffuf, never seen carrying anything.
//
// It matters for more than tidiness. SQLiDetector and xssFuzz both substitute by matching the
// parameter in the URL TEXT, so a parameter that is not physically present matches nothing and the
// tool reports clean. Same word the ffuf baseline uses, so an operator grepping a proxy log for
// rs0n finds everything this framework injected regardless of which tool sent it.
const VectorCanary = FuzzCanaryValue

// vectorRegistry is populated by the per-category files in their init functions, so adding a section
// is adding a file rather than editing this one.
var vectorRegistry []VectorTool

func registerVectorTools(tools ...VectorTool) {
	vectorRegistry = append(vectorRegistry, tools...)
}

// VectorCategories is the display name for each section, in the order the cards appear.
var VectorCategories = []struct {
	Key  string `json:"key"`
	Name string `json:"name"`
}{
	{"xss", "Cross-Site Scripting"},
	{"sqli", "SQL & NoSQL Injection"},
}

// VectorToolsFor returns one section's tools, in registration order.
func VectorToolsFor(category string) []VectorTool {
	var out []VectorTool
	for _, t := range vectorRegistry {
		if t.Category == category {
			out = append(out, t)
		}
	}
	return out
}

// VectorToolByKey looks up one tool. Returns false rather than a zero value, so a bad key from an
// API caller is an error at the edge instead of a scan that runs with no options and no composer.
func VectorToolByKey(key string) (VectorTool, bool) {
	for _, t := range vectorRegistry {
		if t.Key == key {
			return t, true
		}
	}
	return VectorTool{}, false
}

// VectorToolCanReach reports whether a tool can actually fuzz an insertion point.
func VectorToolCanReach(tool VectorTool, insertionPoint string) bool {
	for _, p := range tool.InsertionPoints {
		if p == insertionPoint {
			return true
		}
	}
	return false
}

// UnrecognisedVectorOptions reports keys nothing reads. Never silently dropped: an option that is
// stored and ignored is exactly how a caller comes to believe a setting took effect.
func UnrecognisedVectorOptions(tool VectorTool, settings map[string]any) []string {
	var unknown []string
	for key := range settings {
		if _, ok := tool.Options[key]; !ok {
			unknown = append(unknown, key)
		}
	}
	sort.Strings(unknown)
	return unknown
}

// vectorSettingEngaged reports whether a setting is actually in force.
//
// A switch is engaged when it is true. A setting that carries a VALUE is engaged when it has any
// value at all, which is the whole reason this function exists: dalfox's userAgent blinds the entire
// tool, and truthySetting("Mozilla/5.0 ...") is false, so a value-carrying blinder was invisible to
// the one check whose job is to catch blinders.
func vectorSettingEngaged(tool VectorTool, key string, value any) bool {
	switch tool.Options[key].Kind {
	case "string", "csv", "path", "enum":
		return strings.TrimSpace(stringifySetting(value)) != ""
	default:
		return truthySetting(value)
	}
}

// VectorBlindedPoints reports which insertion points a settings map would silently make untestable.
func VectorBlindedPoints(toolKey string, settings map[string]any) map[string][]string {
	tool, ok := VectorToolByKey(toolKey)
	if !ok || tool.Blinding == nil {
		return nil
	}
	blinded := map[string][]string{}
	for key, point := range tool.Blinding {
		if !vectorSettingEngaged(tool, key, settings[key]) {
			continue
		}
		blinded[point] = append(blinded[point], key)
	}
	for k := range blinded {
		sort.Strings(blinded[k])
	}
	return blinded
}
